SONARJAVA-2728 : Improve secondary issue reporting for S2234

olcbean opened the pull request:
Report secondary location only for the misplace parameters for S2234
( Parameters should be passed in the correct order )

m-g-sonar commented:
hey @olcbean,

I think you are only halfway the fix we were expecting.  
From what I can see, we are adding a "Formal/misplaced parameters" to all of the parameters in the method declaration, while in fact we would expect to have secondaries only on the one which are actually wrongly placed, accordingly to the method invocation. So for instance in the following case:

```java
void foo(
  int a,        //  Misplaced parameter
  int b,        //  parameter ok -> no message
  int c         //  Misplaced parameter       
) { /* ... */}

void bar(int a, int b, int c) {
  foo(c, b, a);
}
```

Regarding testing, there is currently no way in our `JavaCheckVerifier` to test the messages associated to each secondaries... However you can test the **lines** associated to these secondaries. If you spread definition of parameters in multiple lines, you will then be able to test it by adding the `secondaries` part of the issue classifier. Probably using the something similar as the following (assuming `int a` and `int b` would have been defined on line 4 and 6):

```
  foo(c, b, a); // Noncompliant [[secondaries=4,6]] {{bla bla bla}}
```

m-g-sonar commented:
Forgot to say it, but of course, thanks for your contribution as well! :)

olcbean commented:
:facepalm: I totally misread this one, sorry.

@m-g-sonar thanks for the detailed explanation and the example!

I just pushed another commit to address the task at hand and corrected the description accordingly.

Please let me know what you think!

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/checks/MethodParametersOrderCheck.java:
```diff
@@ -68,13 +68,14 @@ public void visitNode(Tree tree) {
+    if (matchingNames(formalParameterList, argumentsList)) {
+      List<VariableTree> matchingTypesWrongOrder = matchingTypesWrongOrder(formalParameterList, argumentsList);
+      if (!matchingTypesWrongOrder.isEmpty()) {
+        List<JavaFileScannerContext.Location> flow = matchingTypesWrongOrder.stream().map(param -> new JavaFileScannerContext.Location("Formal Parameters", param))
```
As this was also causing some trouble, I would still go for the change you previously did: replacing the strange `Formal parameters` by `Misplaced parameter`.

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/checks/MethodParametersOrderCheck.java:
```diff
@@ -98,28 +99,32 @@ public static boolean allUnique(List<String> argListNames) {
       Type argType = argument.symbolType();
       if (!formalType.is(argType.fullyQualifiedName()) || formalType.isUnknown() || argType.isUnknown()) {
-        return false;
+        return new ArrayList<>();
```
I would prefer to use `Collections.emptyList()` here to have the returned collection unmodifiable by other methods.

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/checks/MethodParametersOrderCheck.java:
```diff
@@ -98,28 +99,32 @@ public static boolean allUnique(List<String> argListNames) {
+    if (misplacedParameters.size() >= 2) {
+      return misplacedParameters;
+    }
+    return new ArrayList<>();
```
I would prefer to use `Collections.emptyList()` as well.

olcbean commented:
@m-g-sonar thanks for the review! I just pushed a commit addressing your comments. 

Can you let me know if you prefer the commits be squashed before you open the 'verify' PR.

m-g-sonar commented:
Thx for the update @olcbean . No need to squash everything, I'll do it from here when merging, and I prefer to have clear distinct commits history for review.

m-g-sonar commented:
Merged! Thanks for your contribution!
